sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeOneRowRelationSubquerySuite.scala

```diff
@@ -0,0 +1,161 @@
+  }
+
+  test("Optimize scalar subquery with multiple projects") {
+    // SELECT (SELECT a1 + b1 FROM (SELECT a AS a1, b AS b1)) FROM t1
+    val inner = t0.select('a.as("a1"), 'b.as("b1")).select(('a1 + 'b1).as("c"))
+    val query = t1.select(ScalarSubquery(inner).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    val correctAnswer = Project(Alias(Alias(a + b, "c")(), "sub")() :: Nil, t1)
+    comparePlans(optimized, correctAnswer)
+  }
+
+  test("Optimize lateral subquery with multiple projects") {
+    Seq(Inner, LeftOuter, Cross).foreach { joinType =>
+      val inner = t0.select('a.as("a1"), 'b.as("b1"))
+        .select(('a1 + 'b1).as("c1"), ('a1 - 'b1).as("c2"))
+      val query = t1.lateralJoin(inner, joinType, None)
+      val optimized = Optimize.execute(query.analyze)
+      val correctAnswer = t1.select('a, 'b, ('a + 'b).as("c1"), ('a - 'b).as("c2"))
+      comparePlans(optimized, correctAnswer.analyze)
+    }
+  }
+
+  test("Optimize subquery with nested correlated subqueries") {
+    // SELECT (SELECT (SELECT b) FROM (SELECT a AS b)) FROM t1
+    val inner = t0.select('a.as("b")).select(ScalarSubquery(t0.select('b)).as("s"))
+    val query = t1.select(ScalarSubquery(inner).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    val correctAnswer = Project(Alias(Alias(a, "s")(), "sub")() :: Nil, t1)
+    comparePlans(optimized, correctAnswer)
+  }
+
+  test("Batch should be idempotent") {
+    // SELECT (SELECT 1 WHERE a = a + 1) FROM t1
+    val inner = t0.select(1).where('a === 'a + 1)
+    val query = t1.select(ScalarSubquery(inner).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    val doubleOptimized = Optimize.execute(optimized)
+    comparePlans(optimized, doubleOptimized, checkAnalysis = false)
+  }
+
+  test("Should not optimize scalar subquery with operators other than project") {
+    // SELECT (SELECT a AS a1 WHERE a = 1) FROM t1
+    val inner = t0.where('a === 1).select('a.as("a1"))
+    val query = t1.select(ScalarSubquery(inner).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    assertHasDomainJoin(optimized)
+  }
+
+  test("Should not optimize subquery with non-deterministic expressions") {
+    // SELECT (SELECT r FROM (SELECT a + rand() AS r)) FROM t1
+    val inner = t0.select(('a + rand(0)).as("r")).select('r)
+    val query = t1.select(ScalarSubquery(inner).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    assertHasDomainJoin(optimized)
+  }
+
+  test("Should not optimize lateral join with non-empty join conditions") {
+    Seq(Inner, LeftOuter).foreach { joinType =>
+      // SELECT * FROM t1 JOIN LATERAL (SELECT a AS a1, b AS b1) ON a = b1
+      val query = t1.lateralJoin(t0.select('a.as("a1"), 'b.as("b1")), joinType, Some('a === 'b1))
+      val optimized = Optimize.execute(query.analyze)
+      assertHasDomainJoin(optimized)
+    }
+  }
+
+  test("Should not optimize subquery with nested subqueries") {
```

Comment:
I think we do support nested subqueries, the problem here is `WHERE a = 1`?

Reply:
Yes. The test title is a bit confusing. Will update.